### DISCUSSION TITLE
Adding aria-label to (DivIcon) marker?

### DISCUSSION THREAD

**Discussion by peteruithoven** [Leaflet/Leaflet#9388] [Q&A]
Seeing the custom html of a marker using a DivIcon will always be nested inside the div, why can't we supply an `aria-label` to the [Marker](https://leafletjs.com/reference.html#marker)? Contrary, when supplying a normal image based icon the `alt` can be supplied there. 

Seeing that this outer div gets attributes like `tabindex="0"` and `role="button" ` that seems like the most logical place for a `aria-label`? 

In some places people recommend adding a `aria-label` inside the custom html, but this seems like a workaround? 

I did notice, when the marker has a tooltip, this outer div gets a `aria-describedby` that refers to the tooltip.

**Comment by peteruithoven** [comment]
To add confusion, https://pagespeed.web.dev seems to exclude the `aria-describedby` attribute when the tooltip isn't permanently visible. The browser debug tools can't find the tooltip `aria-describedby` points to unless it's permanently visible. 

When I do make the tooltips permanently visible the `aria-describedby`'s do show up on https://pagespeed.web.dev but it's still saying  "button, link, and menuitem elements do not have accessible names.".

After [reading more](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) I assume, that beside these aria descriptions we should still add some kind of aria label. Either a `aria-label` or [`aria-labelledby`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby).

**Reply by Malvoz** [reply]
If a div icon does not have a tooltip, and no custom HTML, you'll indeed have a button with a missing accessible name, this is to be expected.

In [debug/map/tooltip.html](https://github.com/Leaflet/Leaflet/blob/main/debug/map/tooltip.html) tooltips are conveyed correctly using Narrator, and nothing looks strange in the accessibility tree:

![tooltip-aria](https://github.com/user-attachments/assets/9293e095-6c11-457a-9252-58a59f3dd868)

And all accessibility tests are passed in the Lighthouse audit (except for the document missing the `lang` attribute):

![tooltip-aria-lighthouse](https://github.com/user-attachments/assets/03a4511e-6137-4213-a461-57ac5f9dc1f4)

**Reply by peteruithoven** [reply]
> If a div icon does not have a tooltip, and no custom HTML, you'll indeed have a button with a missing accessible name, this is to be expected.

Well, I disagree or I don't understand you. Buttons need labels. Even when there is no tooltip visible. I think we can agree on that?

**Comment by Malvoz** [comment]
> Seeing that this outer div gets attributes like `tabindex="0"` and `role="button"`  that seems like the most logical place for a `aria-label`?

> In some places people recommend adding a `aria-label` inside the custom html, but this seems like a workaround?

While I may agree, that recommendation is a valid method as the [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name) can be derived from child content of the button (whether text content, `aria-label`/`title`, etc.).

**Reply by peteruithoven** [reply]
I just want to clarify that I find it a workaround because custom html is needed. Making something we probably want to encourage (providing accessible names) harder than it should be.
